### PR TITLE
Add figshare support

### PR DESCRIPTION
**Pull request by sanjayankur31**
adds figshare support for figshare and the ucl instance

the only thing I haven't been able to test properly is if the resource is being downloaded correctly---still seeing the spinner here in the resource pane (but that was supposed to be fixed in #368) creation of the repo, and creation of workspaces from the repo seems to be fine.

Note that this touches a few lines of the editrepo dialog component, so may conflict with https://github.com/OpenSourceBrain/OSBv2/pull/434

### CODE REVIEW

**Review by filippomc — APPROVED** [review]
I'm glad to see from the clean code that it was quite smooth to implement it :) Tested locally and worked, with the exception of an error when I try to create the repository without specifying a tag in the form:

```
[workspaces]
[workspaces]2022-04-01 12:34:53,566 [ERROR] server.handle_exception: 'NoneType' object has no attribute 'append'
[workspaces]Traceback (most recent call last):
[workspaces]  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
[workspaces]    rv = self.dispatch_request()
[workspaces]  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
[workspaces]    return self.view_functions[rule.endpoint](**req.view_args)
[workspaces]  File "/usr/local/lib/python3.7/site-packages/connexion/decorators/decorator.py", line 48, in wrapper
[workspaces]    response = function(request)
[workspaces]  File "/usr/local/lib/python3.7/site-packages/connexion/decorators/security.py", line 327, in wrapper
[workspaces]    return function(request)
[workspaces]  File "/usr/local/lib/python3.7/site-packages/connexion/decorators/uri_parsing.py", line 144, in wrapper
[workspaces]    response = function(request)
[workspaces]  File "/usr/local/lib/python3.7/site-packages/connexion/decorators/validation.py", line 184, in wrapper
[workspaces]    response = function(request)
[workspaces]  File "/usr/local/lib/python3.7/site-packages/connexion/decorators/parameter.py", line 121, in wrapper
[workspaces]    return function(**kwargs)
[workspaces]  File "/usr/src/app/workspaces/views/base_model_view.py", line 44, in post
[workspaces]    obj = self.service.post(body).to_dict()
[workspaces]  File "/libraries/cloudharness-common/cloudharness/events/decorators.py", line 19, in wrapper
[workspaces]    result = func(self, *args, **kwargs)
[workspaces]  File "/usr/src/app/workspaces/service/model_service.py", line 312, in post
[workspaces]    osbrepository.tags.append(Tag(tag=tag))
[workspaces]AttributeError: 'NoneType' object has no attribute 'append'
```

**Review by filippomc — CHANGES_REQUESTED** [review]
sorry approved by mistake, probably we want to fix the error I mentioned

**Comment by sanjayankur31** [comment]
I saw that too, but since it touches the tag code that I've also modified in #434 I didn't want to touch it here---it'll cause a conflict. What's the best way of going forward here? Should we merge both and then see if the issue persists?

**Comment by filippomc** [comment]
> I saw that too, but since it touches the tag code that I've also modified in #434 I didn't want to touch it here---it'll cause a conflict. What's the best way of going forward here? Should we merge both and then see if the issue persists?

OK let's merge #434 first

File: applications/workspaces/server/workspaces/service/osbrepository/adapters/figshareadapter.py [hunk 1]
```diff
@@ -1,34 +1,133 @@
...
         name = name if name != "/" else self.osbrepository.name
-        folder = self.osbrepository.name + path.replace(self.download_base_url + "branches", "")
-        folder = folder[: folder.rfind("/")]
-        # do something...
+        # no file tree in FigShare
+        folder = self.osbrepository.name
+        # username / password are optional and future usage,
+        # e.g. for accessing non public repos
+        return workflow.create_copy_task(
+            image_name="workspaces-figshare-copy",
```
[the remark this answers is not in the document]
Comment (sanjayankur31) [reply applications/workspaces/server/workspaces/service/osbrepository/adapters/figshareadapter.py:126]: Yeh, there's a public download url so this should be usable. Where does one see what `workflows-extract-download` does btw?

Also: why are we using `svn export` in `github-copy`? I thought that with public repos, we'd just be using `curl` or `wget` to get files from their `raw` urls :thinking:

[the remark this answers is not in the document]
Comment (sanjayankur31) [reply applications/workspaces/server/workspaces/service/osbrepository/adapters/figshareadapter.py:126]: found `extract-download` in the cloudharness src

[the remark this answers is not in the document]
Comment (sanjayankur31) [reply applications/workspaces/server/workspaces/service/osbrepository/adapters/figshareadapter.py]: Updated this now.

**Review by filippomc — APPROVED** [review]

**Comment by sanjayankur31** [comment]
> Works great from my tests. As a related thought, we might need to specify something to tell the user that only dataset repository links are valid. It may be not as straightforward as it is for github and dandi to make the mental link.

The way it's implemented now, even links like this one will work (because the regex doesn't include `dataset` in it):

https://figshare.com/articles/presentation/Presentation_1_NWB_Query_Engines_Tools_to_Search_Data_Stored_in_Neurodata_Without_Borders_Format_pdf/12962813

Do we want to limit to `dataset`s only? We'll only need to tweak the regex I expect.

**Comment by filippomc** [comment]
I don't know honestly what's the best approach as I don't know much what we can expect from Figshare. Just shared a general thought about having more variety than the other two repositories, as it may not be the case that non-datasets are valid osb repositories. I'm for taking this in mind and check for feedback as we go with actual users.

**Pull request merged**